hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECKeyOutputStream.java

```diff
@@ -0,0 +1,664 @@
+      return this;
+    }
+
+    public Builder setXceiverClientManager(XceiverClientFactory manager) {
+      this.xceiverManager = manager;
+      return this;
+    }
+
+    public Builder setOmClient(OzoneManagerProtocol client) {
+      this.omClient = client;
+      return this;
+    }
+
+    public Builder setChunkSize(int size) {
+      this.chunkSize = size;
+      return this;
+    }
+
+    public Builder setRequestID(String id) {
+      this.requestID = id;
+      return this;
+    }
+
+    public Builder setIsMultipartKey(boolean isMultipart) {
+      this.isMultipartKey = isMultipart;
+      return this;
+    }
+
+    public Builder setConfig(OzoneClientConfig config) {
+      this.clientConfig = config;
+      return this;
+    }
+
+    public Builder enableUnsafeByteBufferConversion(boolean enabled) {
+      this.unsafeByteBufferConversion = enabled;
+      return this;
+    }
+
+    public ECKeyOutputStream.Builder setReplicationConfig(
+        ReplicationConfig replConfig) {
+      this.replicationConfig = replConfig;
+      return this;
+    }
+
+    public ECKeyOutputStream build() {
+      return new ECKeyOutputStream(clientConfig, openHandler, xceiverManager,
+          omClient, chunkSize, requestID, replicationConfig, multipartUploadID,
+          multipartNumber, isMultipartKey, unsafeByteBufferConversion);
+    }
+  }
+
+  /**
+   * Verify that the output stream is open. Non blocking; this gives
+   * the last state of the volatile {@link #closed} field.
+   *
+   * @throws IOException if the connection is closed.
+   */
+  private void checkNotClosed() throws IOException {
+    if (closed) {
+      throw new IOException(
+          ": " + FSExceptionMessages.STREAM_IS_CLOSED + " Key: "
+              + blockOutputStreamEntryPool.getKeyName());
+    }
+  }
+
+  private static class ECChunkBuffers {
+    private final ByteBuffer[] buffers;
+    private final int dataBlks;
+    private final int parityBlks;
+    private final int totalBlks;
+    private int cellSize;
+
+    ECChunkBuffers(int cellSize, int numData, int numParity) {
+      this.cellSize = cellSize;
+      this.parityBlks = numParity;
+      this.dataBlks = numData;
+      this.totalBlks = this.dataBlks + this.parityBlks;
+      buffers = new ByteBuffer[this.totalBlks];
+      for (int i = 0; i < buffers.length; i++) {
+        buffers[i] = BUFFER_POOL.getBuffer(false, cellSize);
+        buffers[i].limit(cellSize);
+      }
+    }
+
+    private ByteBuffer[] getBuffers() {
+      return buffers;
+    }
+
+    private int addTo(int i, byte[] b, int off, int len) {
+      final ByteBuffer buf = buffers[i];
+      final int pos = buf.position() + len;
+      Preconditions.checkState(pos <= cellSize);
+      buf.put(b, off, len);
+      return pos;
+    }
+
+    private void clear() {
+      for (int i = 0; i < this.totalBlks; i++) {
+        buffers[i].clear();
+        buffers[i].limit(cellSize);
+      }
+    }
+
+    private void release() {
```

Comment:
I don't see this being called anywhere (Intellij marks it as unused). This means we are never returning the buffers to the pool and hence they cannot be reused. Should this be called when the KeyOutputStream() is closed?

Reply:
done.